### ISSUE TITLE
Crop and Decompress

### ISSUE THREAD

**Issue by tilkb** [libjpeg-turbo/libjpeg-turbo#583]
I am considering extending the tjDecompress2 function to support reading cropped images.
I am considering introducing as a new method like `tjCropDecompress2`. Is this feature going to be accepted?

**Comment by dcommander** [comment]
When you say "reading cropped images", do you mean using the partial image decompression API (`jpeg_crop_scanlines()` and `jpeg_skip_scanlines()`?)  If so, then I intend to implement that as part of the TurboVNC 3 API overhaul (#517).  The idea behind that overhaul is to make the TurboJPEG API more of a stateful API rather than a stateless API, so the configuration of each operation (including source/destination images, cropping region, pixel format, algorithms, etc.) will be stored in the opaque instance struct and set using configuration functions.  Thus, the same decompress function could be used regardless of the configuration.  Doing partial image decompression would look something like:

```
tjregion jpegDims, cropRegion;

if (loadJPEGFromFile) {
  if (tj3SetJPEGFile(tjInstance, jpegFileName) < 0) {
    printf("%s (type %d)\n", tj3GetErrorStr(), tj3GetErrorCode());  exit(1);
  }
} else {
  if (tj3SetJPEGBuffer(tjInstance, jpegPtr, jpegSize) < 0) {
    printf("%s (type %d)\n", tj3GetErrorStr(), tj3GetErrorCode());  exit(1);
  }
}
if (tj3GetJPEGDimensions(tjInstance, &jpegDims) < 0) {
  printf("%s (type %d)\n", tj3GetErrorStr(), tj3GetErrorCode());  exit(1);
}
cropRegion = jpegDims.
if (cropRegion.width > 200 && cropRegion.height > 200) {
  cropRegion.width -= 100;
  cropRegion.height -= 100;
  cropRegion.x = cropRegion.y = 100;
  if (tj3SetCropRegion(tjInstance, &cropRegion) < 0) {
    printf("%s (type %d)\n", tj3GetErrorStr(), tj3GetErrorCode());  exit(1);
  }
}
if (saveOutputImage) {
  if (tj3SetOutputFile(tjInstance, outputFileName, TJFORMAT_BMP) < 0) {
    printf("%s (type %d)\n", tj3GetErrorStr(), tj3GetErrorCode());  exit(1);
  }
} else {
  if (tj3SetOutputBuffer(tjInstance, bufPtr, pitch, pixelFormat) < 0) {
    printf("%s (type %d)\n", tj3GetErrorStr(), tj3GetErrorCode());  exit(1);
  }
}
if (tj3Decompress(tjInstance) < 0) {
  printf("%s (type %d)\n", tj3GetErrorStr(), tj3GetErrorCode());  exit(1);
}
```

**Comment by dcommander** [comment]
So to answer your question, no.  It doesn't make sense to extend the existing TurboJPEG API with new functions.

**Comment by dcommander** [comment]
@tilkb Please answer my question regarding whether I properly understood your request.  That will determine whether my existing plan for the TurboJPEG v3 API will accommodate your desired workflow.

**Comment by tilkb** [comment]
@dcommander, sorry for the late answer. Yes partial decompression with `jpeg_crop_scanlines()` and `jpeg_skip_scanlines()` is what I am looking for. Thanks for the information about v3 API.

**Comment by dcommander** [comment]
Subsumed by #517

**Comment by dcommander** [comment]
This is implemented in the TurboJPEG 3 API overhaul (fc01f4673b71c0b833c59c21e8c4478a9c4bcf21), using a new function (`tj3SetCroppingRegion()`) that enables partial image decompression for subsequent decompression operations.  Note that partial image decompression can only be used when decompressing to a packed-pixel image, not when decompressing to a planar YUV image.
